src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs

```diff
@@ -141,11 +141,15 @@ private bool TryRewriteTagHelper(Block tagBlock, RewritingContext context)
             {
                 // We're in an end tag helper block.
 
-                var tagNameScope = _tagStack.Count > 0 ? _tagStack.Peek().TagName : string.Empty;
+                var parentTagHelper = _tagStack.Count > 0 ? _tagStack.Peek() : null;
 
                 // Validate that our end tag helper matches the currently scoped tag helper, if not we
                 // need to error.
-                if (tagNameScope.Equals(tagName, StringComparison.OrdinalIgnoreCase))
+                if (parentTagHelper != null &&
+                    string.Equals(
+                        parentTagHelper.Prefix + parentTagHelper.TagName,
```

Comment:
do not do another concatenation.  something is wrong with the information available to this method.  probably best to simply have `parentTagHelper.TagName` contain the full name that was matched for the start tag.